testing/duck/testbindable.go

```diff
@@ -0,0 +1,169 @@
+*/
+
+package duck
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	duckv1alpha1 "knative.dev/pkg/apis/duck/v1alpha1"
+	"knative.dev/pkg/tracker"
+	"knative.dev/pkg/webhook/psbinding"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// TestBindable is a simple resource that's compatible with our webhook
+type TestBindable struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   TestBindableSpec   `json:"spec,omitempty"`
+	Status TestBindableStatus `json:"status,omitempty"`
+}
+
+// Check that TestBindable may be validated and defaulted.
+var _ apis.Listable = (*TestBindable)(nil)
+var _ psbinding.Bindable = (*TestBindable)(nil)
+
+// TestBindableSpec represents test resource spec.
+type TestBindableSpec struct {
+	duckv1alpha1.BindingSpec `json:",inline"`
+
+	Foo string `json:"foo"`
+}
+
+// TestBindableStatus represents the status of our test binding.
+type TestBindableStatus struct {
+	duckv1.Status `json:",inline"`
+}
+
+// GetGroupVersionKind returns the GroupVersionKind.
+func (r *TestBindable) GetGroupVersionKind() schema.GroupVersionKind {
+	return SchemeGroupVersion.WithKind("TestBindable")
+}
+
+// GetUntypedSpec returns the spec of the resource.
+func (r *TestBindable) GetUntypedSpec() interface{} {
+	return r.Spec
+}
+
+// GetListType implements apis.Listable
+func (r *TestBindable) GetListType() runtime.Object {
+	return &TestBindableList{}
+}
+
+// GetSubject implements psbinding.Bindable
+func (fb *TestBindable) GetSubject() tracker.Reference {
+	return fb.Spec.Subject
+}
+
+// GetBindingStatus implements psbinding.Bindable
+func (fb *TestBindable) GetBindingStatus() psbinding.BindableStatus {
+	return &fb.Status
+}
+
+// SetObservedGeneration implements psbinding.BindableStatus
+func (fbs *TestBindableStatus) SetObservedGeneration(gen int64) {
+	fbs.ObservedGeneration = gen
+}
+
+// Do implements psbinding.Bindable
+func (fb *TestBindable) Do(ctx context.Context, ps *duckv1.WithPod) {
+	// First undo so that we can just unconditionally append below.
+	fb.Undo(ctx, ps)
+
+	spec := ps.Spec.Template.Spec
+	for i := range spec.InitContainers {
+		spec.InitContainers[i].Env = append(spec.InitContainers[i].Env, corev1.EnvVar{
+			Name:  "FOO",
+			Value: fb.Spec.Foo,
+		})
+	}
+	for i := range spec.Containers {
+		spec.Containers[i].Env = append(spec.Containers[i].Env, corev1.EnvVar{
+			Name:  "FOO",
+			Value: fb.Spec.Foo,
+		})
+	}
+}
+
+func (fb *TestBindable) Undo(ctx context.Context, ps *duckv1.WithPod) {
+	spec := ps.Spec.Template.Spec
+	for i, c := range spec.InitContainers {
+		for j, ev := range c.Env {
+			if ev.Name == "FOO" {
+				spec.InitContainers[i].Env = append(spec.InitContainers[i].Env[:j], spec.InitContainers[i].Env[j+1:]...)
```

Comment:
Technically that changes ordering, which can be semantic since I don't believe Kubernetes validates away:
```
env:
- name: FOO
  value: BAR
- name: FOO
  value: BAZ
```